transports/tcp/src/if_task.rs

```diff
@@ -0,0 +1,151 @@
+use futures::{
+    channel::{mpsc, oneshot},
+    prelude::*,
+    select,
+};
+use if_watch::{IfEvent, IfWatcher, IpNet};
+use std::{io, net::SocketAddr};
+
+#[derive(Clone, Copy, Debug)]
+pub enum IfTaskEvent {
+    NewAddress(SocketAddr),
+    AddressExpired(SocketAddr),
+}
+
+#[derive(Debug)]
+enum Command {
+    RegisterListener(SocketAddr, mpsc::UnboundedSender<IfTaskEvent>),
+    PortReuseSocket(SocketAddr, oneshot::Sender<Option<SocketAddr>>),
+}
+
+#[derive(Clone, Debug)]
+pub struct IfTaskHandle {
+    tx: mpsc::UnboundedSender<Command>,
+}
+
+impl IfTaskHandle {
+    pub async fn new() -> io::Result<Self> {
+        // channel is unbounded so that `register_listener` doesn't need to be async.
+        let (tx, rx) = mpsc::unbounded();
+        let task = Box::pin(Task::new(rx).await?.spawn());
+        #[cfg(feature = "async-std")]
+        async_std::task::spawn(task);
+        #[cfg(all(not(feature = "async-std"), feature = "tokio"))]
+        tokio::task::spawn(task);
+        #[cfg(all(not(feature = "async-std"), not(feature = "tokio")))]
+        std::thread::spawn(move || futures::executor::block_on(task));
+        Ok(Self { tx })
```

Comment:
This is a bit of a hack, not sure how to spawn tasks in an executor independent way. Because we are sharing a background task between listeners for polling interface changes, `TcpConfig::new()` returns a future, which is also not great.

Reply:
Fixed using agnostik, still waiting for https://github.com/bastion-rs/agnostik/pull/28

Reply:
giving up on this one, there is no adequate solution. new needs to take a spawner as an argument :/